op-node/rollup/driver/origin_selector.go

```diff
@@ -0,0 +1,69 @@
+package driver
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+)
+
+type L1Blocks interface {
+	derive.L1BlockRefByHashFetcher
+	derive.L1BlockRefByNumberFetcher
+}
+
+type L1OriginSelector struct {
+	Cfg                 *rollup.Config
+	Log                 log.Logger
+	L1                  L1Blocks
+	SequencingConfDepth uint64
+}
+
+// FindL1Origin determines what the next L1 Origin should be.
+// The L1 Origin is either the L2 Head's Origin, or the following L1 block
+// if the next L2 block's time is greater than or equal to the L2 Head's Origin.
+func (los *L1OriginSelector) FindL1Origin(ctx context.Context, l1Head eth.L1BlockRef, l2Head eth.L2BlockRef) (eth.L1BlockRef, error) {
```

Comment:
I'm starting to wonder if this should be merged in with `L1 State`. Outside of `syncStatus`, this is the only usage of the `l1Head`.

Reply:
Good point about `l1Head`. The safe/finalized L1 data however is not part of this function, so I think it's better to isolate, to keep tests small and logic more encapsulated.

Reply:
other than sync status (which is nice to have), we don't really use the rest of the L1 State. It's not an immediate change, but I think it might be a possible simplification down the road.